File: 0169-majority-element/0169-majority-element.cpp
class Solution {
public:
    int majorityElement(vector<int>& nums) {
        int majorityCount = nums.size() / 2;
        while (true) {
            int candidate = nums[rand() % nums.size()];
            if (std::ranges::count_if(nums, [&](auto num) { return candidate == num; }) > majorityCount) {
                return candidate;
            }
        }
    }
};

File: 0169-majority-element/0169-majority-element.py
class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        current_candidate = nums[0]
        current_count = 1

        for i in range(1, len(nums)):
            if current_candidate is None:
                current_candidate = nums[i]
                current_count += 1
                continue

            if current_candidate == nums[i]:
                current_count += 1
            else:
                current_count -= 1

            if current_count == 0:
                current_candidate = None

        
        return current_candidate